src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/WebRequestPSCmdlet.Common.cs

```diff
@@ -1312,24 +1323,42 @@ internal virtual HttpResponseMessage GetResponse(HttpClient client, HttpRequestM
 
                 _cancelToken = new CancellationTokenSource();
                 response = client.SendAsync(req, HttpCompletionOption.ResponseHeadersRead, _cancelToken.Token).GetAwaiter().GetResult();
+                
+                bool sessionRedirect = WebSession.MaximumRedirection > 0 || WebSession.MaximumRedirection == -1;
 
-                if (keepAuthorization && IsRedirectCode(response.StatusCode) && response.Headers.Location is not null)
+                if ((keepAuthorization || (PreserveHTTPMethodOnRedirect && sessionRedirect)) && IsRedirectCode(response.StatusCode) && response.Headers.Location is not null)
                 {
                     _cancelToken.Cancel();
                     _cancelToken = null;
 
-                    // if explicit count was provided, reduce it for this redirection.
+                    // If explicit count was provided, reduce it for this redirection.
                     if (WebSession.MaximumRedirection > 0)
                     {
                         WebSession.MaximumRedirection--;
                     }
-                    // For selected redirects that used POST, GET must be used with the
-                    // redirected Location.
-                    // Since GET is the default; POST only occurs when -Method POST is used.
-                    if (Method == WebRequestMethod.Post && IsRedirectToGet(response.StatusCode))
+
+                    // For selected redirects, GET must be used with the redirected Location.
+                    // See https://msdn.microsoft.com/library/system.net.httpstatuscode(v=vs.110).aspx
+                    if (RequestRequiresForceGet(response.StatusCode, Method))
                     {
-                        // See https://msdn.microsoft.com/library/system.net.httpstatuscode(v=vs.110).aspx
-                        Method = WebRequestMethod.Get;
+                        if (PreserveHTTPMethodOnRedirect)
+                        {
+                            switch (response.StatusCode)
```

Comment:
Why do we need the switch if it is already in RequestRequiresForceGet?

Reply:
We no longer need it, after switching to SwitchParameter